kani-compiler/src/kani_middle/stubbing/annotations.rs

```diff
@@ -0,0 +1,105 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+//! This file contains code for extracting stubbing-related attributes.
+
+use rustc_ast::Attribute;
+use rustc_data_structures::fx::FxHashMap;
+use rustc_driver::RunCompiler;
+use rustc_driver::{Callbacks, Compilation};
+use rustc_errors::ErrorGuaranteed;
+use rustc_interface::interface::Compiler;
+use rustc_interface::Queries;
+use rustc_middle::ty::TyCtxt;
+
+use crate::kani_middle::attributes::{extract_path_arguments, partition_kanitool_attributes};
+
+/// Collects the stubs from the harnesses in a crate, running rustc (to
+/// expansion) with the supplied arguments `rustc_args`.
+pub fn collect_stub_mappings(
+    rustc_args: &[String],
+) -> Result<FxHashMap<String, FxHashMap<String, String>>, ErrorGuaranteed> {
+    let mut callbacks = CollectorCallbacks::default();
+    let compiler = RunCompiler::new(rustc_args, &mut callbacks);
+    compiler.run().map(|_| callbacks.stub_mapping)
+}
+
+/// A rustc callback that is used to collect the stub mappings specified for
+/// each harness.
+#[derive(Default)]
+struct CollectorCallbacks {
+    stub_mapping: FxHashMap<String, FxHashMap<String, String>>,
+}
+
+impl Callbacks for CollectorCallbacks {
+    /// The main callback, invoked after the HIR has been created.
+    fn after_expansion<'tcx>(
+        &mut self,
+        _compiler: &Compiler,
+        queries: &'tcx Queries<'tcx>,
+    ) -> Compilation {
+        queries.global_ctxt().unwrap().peek_mut().enter(|tcx| {
+            for item in tcx.hir_crate_items(()).items() {
+                let def_id = item.owner_id.def_id.to_def_id();
+                let (proof, other) = partition_kanitool_attributes(tcx.get_attrs_unchecked(def_id));
+                // Ignore anything that is not a harness
+                if proof.is_empty() {
+                    continue;
+                }
+                let mut stub_pairs = FxHashMap::default();
+                for (name, attr) in other {
+                    if name == "stub" {
+                        update_stub_mapping(tcx, attr, &mut stub_pairs);
+                    }
+                }
+                let harness_name = tcx.def_path_str(def_id);
+                self.stub_mapping.insert(harness_name, stub_pairs);
+            }
+            tcx.sess.abort_if_errors();
```

Comment:
Was expecting this to be in the loop, I'm not sure you want to continue it if errors get spanned for any proof.

Reply:
The nice thing about leaving this after the loop is that you print all errors at once. So users don't find a new error after they fix the first one.

Reply:
Yes, that's my justification for it. I think it would be annoying for the user to have to see errors one at a time.

Reply:
That makes sense, thanks!